File: apk/app/src/main/java/com/example/momentra/data/repository/MomentCreateRepository.kt
package com.example.momentra.data.repository

import com.example.momentra.data.api.ApiClient
import com.example.momentra.data.api.ApiService
import com.example.momentra.data.api.BusinessSetupBlockDto
import com.example.momentra.data.api.CreateMomentBody
import com.example.momentra.data.api.CreateMomentParticipantBody
import com.example.momentra.data.api.CreateMomentResultDto
import com.example.momentra.data.api.GroupInviteDto
import com.example.momentra.data.api.MintGroupInviteBody
import com.example.momentra.data.api.GroupSetupBlockDto
import com.example.momentra.data.api.PersonalSetupBlockDto
import com.example.momentra.data.api.mapHttpFailure
import retrofit2.HttpException
import java.io.IOException
import java.util.TimeZone
import java.util.UUID

class MomentCreateRepository(
    private val api: ApiService = ApiClient.apiService,
) {
    suspend fun createPersonalMoment(
        systemCode: String,
        momentTypeCode: String,
        title: String,
        preferences: Map<String, Any>? = null,
        idempotencyKey: String = UUID.randomUUID().toString(),
    ): Result<CreateMomentResultDto> = runCatching {
        api.createMoment(
            idempotencyKey = idempotencyKey,
            body = CreateMomentBody(
                domainCode = "PERSONAL",
                momentTypeCode = momentTypeCode,
                title = title,
                personalSetup = PersonalSetupBlockDto(systemCode = systemCode, preferences = preferences),
            ),
        ).data
    }.recoverCatching { e -> throw mapCreateError(e) }

    suspend fun createBusinessMoment(
        companyId: String,
        familyCode: String,
        momentTypeCode: String,
        title: String,
        preferences: Map<String, Any>? = null,
        idempotencyKey: String = UUID.randomUUID().toString(),
    ): Result<CreateMomentResultDto> = runCatching {
        api.createMoment(
            idempotencyKey = idempotencyKey,
            body = CreateMomentBody(
                domainCode = "BUSINESS",
                momentTypeCode = momentTypeCode,
                title = title,
                companyId = companyId,
                businessSetup = BusinessSetupBlockDto(familyCode = familyCode, preferences = preferences),
            ),
        ).data
    }.recoverCatching { e -> throw mapCreateError(e) }

    suspend fun createGroupMoment(
        momentTypeCode: String,
        title: String,
        description: String? = null,
        startAt: String? = null,
        endAt: String? = null,
        participants: List<CreateMomentParticipantBody> = emptyList(),
        inviteCode: String? = null,
        customTypeLabel: String? = null,
        groupSetup: GroupSetupBlockDto? = null,
        idempotencyKey: String = UUID.randomUUID().toString(),
    ): Result<CreateMomentResultDto> = runCatching {
        api.createMoment(
            idempotencyKey = idempotencyKey,
            body = CreateMomentBody(
                domainCode = "GROUP",
                momentTypeCode = momentTypeCode,
                title = title,
                description = description,
                startAt = startAt,
                endAt = endAt,
                timezone = TimeZone.getDefault().id,
                customTypeLabel = customTypeLabel,
                participants = participants.ifEmpty { null },
                inviteCode = inviteCode,
                groupSetup = groupSetup,
            ),
        ).data
    }.recoverCatching { e -> throw mapCreateError(e) }

    suspend fun mintGroupInvite(
        title: String,
        momentTypeCode: String,
        idempotencyKey: String = UUID.randomUUID().toString(),
    ): Result<GroupInviteDto> = runCatching {
        api.mintGroupInvite(
            idempotencyKey = idempotencyKey,
            body = MintGroupInviteBody(title = title, momentTypeCode = momentTypeCode),
        ).data
    }.recoverCatching { e -> throw mapCreateError(e) }

    private fun mapCreateError(e: Throwable): Throwable = when (e) {
        is HttpException -> {
            val body = e.response()?.errorBody()?.string()
            val code = Regex("\"code\"\\s*:\\s*\"([^\"]+)\"").find(body.orEmpty())?.groupValues?.getOrNull(1)
            val msg = Regex("\"message\"\\s*:\\s*\"([^\"]+)\"").find(body.orEmpty())?.groupValues?.getOrNull(1)
            mapHttpFailure(e.code(), code, msg)
        }
        is IOException -> com.example.momentra.data.api.ApiResultException.Network(cause = e)
        else -> e
    }
}

File: momentra/momentra/Data/MomentCreateRepository.swift
import Foundation

protocol MomentCreateGateway {
    func createMoment(draftKey: String, body: CreateMomentRequest) async throws -> CreateMomentOutcome
    func createPersonalSetup(
        draftKey: String,
        systemCode: String,
        title: String,
        description: String?,
        momentTypeCode: String,
        preferences: [String: Any]
    ) async throws -> CreateMomentOutcome
    func createBusinessSetup(
        draftKey: String,
        companyId: String,
        familyCode: String,
        title: String,
        description: String?,
        momentTypeCode: String,
        preferences: [String: Any]
    ) async throws -> CreateMomentOutcome
    func createGroupMoment(
        draftKey: String,
        momentTypeCode: String,
        title: String,
        description: String?,
        startAt: String?,
        endAt: String?,
        participants: [CreateMomentParticipantInput],
        inviteCode: String?,
        customTypeLabel: String?,
        groupSetup: CreateMomentRequest.GroupSetupBlock?
    ) async throws -> CreateMomentOutcome
    func mintGroupInvite(draftKey: String, title: String, momentTypeCode: String) async throws -> GroupInvite
    func previewGroupInvite(code: String) async throws -> GroupInvite
    func redeemGroupInvite(code: String) async throws -> RedeemGroupInviteResult
    func mintCompanyInvite(companyId: String, membershipType: String) async throws -> CompanyInvite
    func previewCompanyInvite(code: String) async throws -> CompanyInvite
    func redeemCompanyInvite(code: String) async throws -> RedeemCompanyInviteResult
    func listCompanies() async throws -> [CompanySummary]
}

final class MomentCreateRepository: MomentCreateGateway {
    private let client: APIClient
    private let idempotency: IdempotencyKeyStore

    init(client: APIClient = .shared, idempotency: IdempotencyKeyStore = IdempotencyKeyStore()) {
        self.client = client
        self.idempotency = idempotency
    }

    func createMoment(draftKey: String, body: CreateMomentRequest) async throws -> CreateMomentOutcome {
        try await runCreate(draftKey: draftKey) { key in
            try await client.createMoment(body: body, idempotencyKey: key)
        }
    }

    func createPersonalSetup(
        draftKey: String,
        systemCode: String,
        title: String,
        description: String?,
        momentTypeCode: String,
        preferences: [String: Any]
    ) async throws -> CreateMomentOutcome {
        try await createMoment(
            draftKey: draftKey,
            body: CreateMomentRequest(
                domainCode: "PERSONAL",
                momentTypeCode: momentTypeCode,
                title: title,
                description: description,
                startAt: nil,
                endAt: nil,
                timezone: TimeZone.current.identifier,
                companyId: nil,
                participants: nil,
                personalSetup: .init(
                    systemCode: systemCode,
                    preferences: JSONEncodableValue.map(preferences)
                ),
                businessSetup: nil
            )
        )
    }

    func createBusinessSetup(
        draftKey: String,
        companyId: String,
        familyCode: String,
        title: String,
        description: String?,
        momentTypeCode: String,
        preferences: [String: Any]
    ) async throws -> CreateMomentOutcome {
        try await createMoment(
            draftKey: draftKey,
            body: CreateMomentRequest(
                domainCode: "BUSINESS",
                momentTypeCode: momentTypeCode,
                title: title,
                description: description,
                startAt: nil,
                endAt: nil,
                timezone: TimeZone.current.identifier,
                companyId: companyId,
                participants: nil,
                personalSetup: nil,
                businessSetup: .init(
                    familyCode: familyCode,
                    preferences: JSONEncodableValue.map(preferences)
                )
            )
        )
    }

    func createGroupMoment(
        draftKey: String,
        momentTypeCode: String,
        title: String,
        description: String?,
        startAt: String?,
        endAt: String?,
        participants: [CreateMomentParticipantInput],
        inviteCode: String?,
        customTypeLabel: String? = nil,
        groupSetup: CreateMomentRequest.GroupSetupBlock? = nil
    ) async throws -> CreateMomentOutcome {
        try await createMoment(
            draftKey: draftKey,
            body: CreateMomentRequest(
                domainCode: "GROUP",
                momentTypeCode: momentTypeCode,
                title: title,
                description: description,
                startAt: startAt,
                endAt: endAt,
                timezone: TimeZone.current.identifier,
                customTypeLabel: customTypeLabel,
                companyId: nil,
                participants: participants.isEmpty ? nil : participants,
                inviteCode: inviteCode,
                personalSetup: nil,
                businessSetup: nil,
                groupSetup: groupSetup
            )
        )
    }

    func mintGroupInvite(draftKey: String, title: String, momentTypeCode: String) async throws -> GroupInvite {
        let key = idempotency.keyFor(draftKey: draftKey)
        return try await client.mintGroupInvite(
            title: title,
            momentTypeCode: momentTypeCode,
            idempotencyKey: key
        )
    }

    func redeemGroupInvite(code: String) async throws -> RedeemGroupInviteResult {
        try await client.redeemGroupInvite(code: code, idempotencyKey: UUID().uuidString)
    }

    func previewGroupInvite(code: String) async throws -> GroupInvite {
        try await client.getGroupInvite(code: code)
    }

    func previewCompanyInvite(code: String) async throws -> CompanyInvite {
        try await client.getCompanyInvite(code: code)
    }

    func mintCompanyInvite(companyId: String, membershipType: String) async throws -> CompanyInvite {
        try await client.mintCompanyInvite(
            companyId: companyId,
            membershipType: membershipType,
            idempotencyKey: UUID().uuidString
        )
    }

    func redeemCompanyInvite(code: String) async throws -> RedeemCompanyInviteResult {
        try await client.redeemCompanyInvite(code: code, idempotencyKey: UUID().uuidString)
    }

    func listCompanies() async throws -> [CompanySummary] {
        try await client.listCompanies()
    }

    private func runCreate(
        draftKey: String,
        call: (String) async throws -> CreateMomentAPIResponse
    ) async throws -> CreateMomentOutcome {
        let key = idempotency.keyFor(draftKey: draftKey)
        do {
            let response = try await call(key)
            idempotency.clear(draftKey: draftKey)
            return response.toOutcome()
        } catch {
            throw error
        }
    }
}

private extension CreateMomentAPIResponse {
    func toOutcome() -> CreateMomentOutcome {
        CreateMomentOutcome(
            momentId: result.momentId,
            title: result.title,
            domainCode: result.domainCode,
            status: result.status,
            version: result.version,
            momentTypeCode: result.momentTypeCode,
            setupId: result.setupId,
            projectionHints: projectionHints.map {
                ProjectionHint(projection: $0.projection, action: $0.action ?? "invalidate")
            }
        )
    }
}
